### ISSUE TITLE
expression too large

### ISSUE THREAD

**Issue by marius311** [JuliaDiff/ReverseDiff.jl#52]
I think I'm hitting a limit to the number of variables I can have near 100,000, e.g., 

```
$ julia 
               _
   _       _ _(_)_     |  A fresh approach to technical computing
  (_)     | (_) (_)    |  Documentation: http://docs.julialang.org
   _ _   _| |_  __ _   |  Type "?help" for help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 0.5.0 (2016-09-19 18:14 UTC)
 _/ |\__'_|_|_|\__'_|  |  Official http://julialang.org/ release
|__/                   |  x86_64-pc-linux-gnu

julia> using ReverseDiff: compile_gradient

julia> compile_gradient(x->norm(x), rand(10000))
(::#301) (generic function with 1 method)

julia> compile_gradient(x->norm(x), rand(100000))
ERROR: syntax: expression too large
 in compile(::ReverseDiff.GradientTape{##7#8,ReverseDiff.TrackedArray{Float64,Float64,1,Array{Float64,1},Array{Float64,1}},ReverseDiff.TrackedReal{Float64,Float64,Void}}) at /home/marius/.julia/v0.5/ReverseDiff/src/api/tape.jl:83
 in compile_gradient(::Function, ::Array{Float64,1}, ::Vararg{Array{Float64,1},N}) at /home/marius/.julia/v0.5/ReverseDiff/src/api/tape.jl:104
```

Is there any way to go beyond this or is just impractical?

Actually, where I'd be interested if I can get to is about O(10,000,000)... is that just too crazy to even consider?

**Comment by jrevels** [comment]
Note that the problem here isn't the size of the input per se, but really the size of the generated instruction tape. If we simply define the `norm` instruction, then ReverseDiff wouldn't have to unroll the function into a large number of intermediate instructions - we'd only need the single `norm` instruction no matter the size of the input. The paper linked in #42 covers the implementation ReverseDiff would use for `norm` (as well as some other derivatives we still need primitives for).

More generally, for very large tapes that are impractical to compile, [one can use the non-compiled API functions](http://www.juliadiff.org/ReverseDiff.jl/api/#ReverseDiff.gradient) along with [ReverseDiff's taping API](http://www.juliadiff.org/ReverseDiff.jl/api/#ReverseDiff.GradientTape) so that you don't have to re-record the tape for every execution.

I just realized the gradient example doesn't cover this, I should definitely add that in.

**Comment by jrevels** [comment]
> Any ideas to squeeze just another factor of 10 without having to go to a bigger RAM machine?

Depending on what your actual objective function looks like, you may be able to sprinkle `@forward` in some clever places to cut down on the overall number of instructions. It's hard to say without seeing the code though.

> is there a way I can input them in by hand to ReverseDiff uses them?

Yup, but I haven't exposed a user-facing API for defining new instructions yet so it'd probably be pretty cumbersome for anybody who doesn't already know ReverseDiff's internals (#15 exists to track that feature). If you wanted to play around with it, though, I'd be willing to help out. 

Basically, to support a new function instruction, you have to define three things: how to "record" the function to a tape instruction, how to execute it during the forward pass, and how to execute it during the backwards pass. The [`inv` definition here](https://github.com/JuliaDiff/ReverseDiff.jl/blob/master/src/derivatives/linalg/special.jl#L39) is probably the most straightforward example at the moment.
